File: bms/go/geo/geo_s2_CellIndexRangeIterator_Begin/geo_s2_CellIndexRangeIterator_Begin.go
package test

const (
	geo_s2_cellIndexDoneContents = -1
)

func (c *geo_s2_CellIndexRangeIterator) Begin() {
	c.pos = 0
	for c.nonEmpty && c.IsEmpty() && !c.Done() {
		c.pos++
	}
}

func (c *geo_s2_CellIndexRangeIterator) Done() bool {
	return c.pos >= len(c.rangeNodes)-1
}

func (c *geo_s2_CellIndexRangeIterator) IsEmpty() bool {
	return c.rangeNodes[c.pos].contents == geo_s2_cellIndexDoneContents
}

type geo_s2_CellIndexRangeIterator struct {
	rangeNodes []geo_s2_rangeNode
	pos        int
	nonEmpty   bool
}

type geo_s2_rangeNode struct {
	// First leaf cell contained by this range.
	contents int32 // Contents of this node (an index within the cell tree).
}

File: translations_only/fail/go/geo/geo_s2_CellIndexRangeIterator_Begin/gemini/Hinted/geo_s2_CellIndexRangeIterator_Begin.rs

const GEO_S2_CELL_INDEX_DONE_CONTENTS: i32 = -1;

pub struct GeoS2CellIndexRangeIterator {
    range_nodes: Vec<GeoS2RangeNode>,
    pos: usize,
    non_empty: bool,
}

impl GeoS2CellIndexRangeIterator {
    pub fn begin(&mut self) {
        self.pos = 0;
        while self.non_empty && self.is_empty() && !self.done() {
            self.pos += 1;
        }
    }

    pub fn done(&self) -> bool {
        self.pos >= self.range_nodes.len() - 1
    }

    pub fn is_empty(&self) -> bool {
        self.range_nodes[self.pos].contents == GEO_S2_CELL_INDEX_DONE_CONTENTS
    }
}

pub struct GeoS2RangeNode {
    contents: i32,
}
